ops/charm.py

```diff
@@ -1297,16 +1338,82 @@ def __init__(self, raw: Optional[Dict[str, Any]] = None,
                          for name, storage in raw_.get('storage', {}).items()}
         self.resources = {name: ResourceMeta(name, res)
                           for name, res in raw_.get('resources', {}).items()}
+        self.devices = {name: DeviceMeta.from_dict(device)
+                        for name, device in raw_.get('devices', {}).items()}
         self.payloads = {name: PayloadMeta(name, payload)
                          for name, payload in raw_.get('payloads', {}).items()}
         self.extra_bindings = raw_.get('extra-bindings', {})
         self.actions = {name: ActionMeta(name, action) for name, action in actions_raw_.items()}
-        # This is taken from Charm Metadata v2, but only the "containers" and
-        # "containers.name" fields that we need right now for Pebble. See:
-        # https://discourse.charmhub.io/t/charm-metadata-v2/3674
         self.containers = {name: ContainerMeta(name, container)
                            for name, container in raw_.get('containers', {}).items()}
 
+    def _load_links(self, raw: Dict[str, Any]):
+        websites = raw.get('website', [])
+        if not websites and 'links' in raw:
+            # When running tests, this might be loading from charmcraft.yaml.
+            websites = raw['links'].get('website', [])
+        # In YAML, this can be a single string, or a list of strings.
+        if isinstance(websites, str):
+            websites = [websites]
+        sources = raw.get('source', [])
+        if not sources and 'links' in raw:
+            # When running tests, this might be loading from charmcraft.yaml.
+            sources = raw['links'].get('source', [])
+        # In YAML, this can be a single string, or a list of strings.
+        if isinstance(sources, str):
+            sources = [sources]
+        issues = raw.get('issues', [])
+        if not issues and 'links' in raw:
+            # When running tests, this might be loading from charmcraft.yaml.
+            issues = raw['links'].get('issues', [])
+        # In YAML, this can be a single string, or a list of strings.
+        if isinstance(issues, str):
+            issues = [issues]
+        documentation = raw.get('documentation')
+        if documentation is None:
+            # When running tests, this might be loading from charmcraft.yaml.
+            documentation = raw.get('links', {}).get('documentation')
+        self.links = MetadataLinks(
+            websites=websites,
+            sources=sources,
+            issues=issues,
+            documentation=documentation,
+        )
+
+    def _assumes_min_juju_version(self) -> Optional[str]:
```

Comment:
@benhoyt what this does is provide a compatibility layer so that if you have `assumes: xxx` in your yaml, your charm can still use `meta.min_juju_version` and it'll pick up any minimum version specified in the `assumes` block.

This code works in my quick manual testing, but I'd need to write a bunch of tests to properly validate it, check edge cases, etc.

I'm wondering now if it's actually worth doing this. Maybe in this situation the answer is just for the charmer to adjust their code to use `meta.assumes` instead? They would then need to have code similar to this themselves, but it might be that they need to actually use the maximum or other things rather than just the minimum, and maybe if that's common then down the track it would be better to have some helpers in ops.jujuversion or something similar.

Thoughts?

Reply:
Yeah, it's a fair bit of new code/logic. I think we should keep the metadata dumb. If the charmer really needs to do this they can, but seems unlikely they'll need to in 99% of cases.